### PR TITLE
RFC: Move iterative eigensolvers to the stdlib

### PR DESCRIPTION
**Pull request by ararslan**
This PR moves `eigs` and `svds` into an Arnoldi package in the stdlib. This was very self-contained and was a pretty easy move. Julia's binary dependency on ARPACK is now only for this package, which can eventually be moved out entirely. ~~thereby eliminating a GPL-licensed binary dependency.~~ (Just kidding, ARPACK is 3-clause BSD.)

~~There might be a couple of uses of `eigs` that I missed, hence WIP.~~ Done.

### CODE REVIEW

File: test/choosetests.jl [hunk 1]
```diff
@@ -127,9 +127,6 @@ function choosetests(choices = [])
                    "linalg/generic", "linalg/uniformscaling", "linalg/lq",
                    "linalg/hessenberg", "linalg/rowvector", "linalg/conjarray",
                    "linalg/blas"]
-    if Base.USE_GPL_LIBS
```
Comment (ararslan) [inline test/choosetests.jl]: This is confusing to me. ARPACK is not GPL-licensed, so why did we only test the Arnoldi stuff if we were using GPL libraries?

Comment (StefanKarpinski) [reply test/choosetests.jl]: Guessing: maybe because it depends on SuiteSparse which is partly GPL?

Comment (ararslan) [reply test/choosetests.jl]: That would make sense. I guess I should modify the stdlib tests then to only run Arnoldi when we're okay with using GPL libraries?

Comment (StefanKarpinski) [reply test/choosetests.jl]: @ViralBShah ^

Comment (andreasnoack) [reply test/choosetests.jl]: So ARPACK as such doesn't depend on anything in SuiteSparse but the wrappers use factorizations and when the input is sparse, which is the case in some of the tests, then the wrappers use sparse factorizations from SuiteSparse.

**Comment by andreasnoack** [comment]
So the way to think of `arpack.jl` and `arnoldi.jl` is that `arpack.jl` is like `blas.jl` and `lapack.jl`, i.e. fairly low-level wrappers of Fortran subroutines while `arnoldi.jl` is like `cholesky.jl`, `qr.jl`, and `lu.jl`, i.e. higher level functionality that utilizes the wrappers in `arpack.jl`.

We should consider a more recognizable name than `Arnoldi` for the module because many of the users of `eigs` will have no idea what Arnoldi is. Although these methods are (almost) agnostic to the storage format, it might be an idea to call it `SparseEigen` or something like that because they are most often used with sparse matrices. Any thoughts @Sacha0 and @fredrikekre?

**Comment by fredrikekre** [comment]
> We should consider a more recognizable name than `Arnoldi`

Yes, I also think that `Arnoldi` is a bad name. I suggested simply `ARPACK` in `slack/linalg`, but as @ararslan pointed out we are not wrapping all of `ARPACK` (but perhaps that will change in the future, and then ARPACK would be an appropriate name). Brainstorming some other names: `EigenIterations`, `IterativeEigen`, `IterativeEigenSolvers`. I kind of like `EigenIterations`.

**Comment by stevengj** [comment]
We should also avoid a name that sounds like it is about iterators. Maybe EigSubspace?

**Comment by Sacha0** [comment]
A better name than `Arnoldi` would be great. `@fredrikekre`'s `IterativeEigenSolvers` strikes me as the most descriptive and accurate from those above. Best!

**Comment by ararslan** [comment]
IterativeEigenSolvers seems fine, though it's quite similar to [JuliaMath/IterativeSolvers](https://github.com/JuliaMath/IterativeSolvers.jl) (not sure whether that's good or bad). According to Andreas, we do wrap most of ARPACK, but I suppose I would expect a package with a name like ARPACK to be a lower level wrapper over the library rather than providing higher level APIs as we do in this package.

I'm pretty indifferent toward what this is called, so once a consensus is reached I'll just move forward with the agreed upon name.

**Comment by Sacha0** [comment]
> I would expect a package with a name like ARPACK to be a lower level wrapper over the library rather than providing higher level APIs as we do in this package.

Likewise, i.e. I would expect the packaged `IterativeEigenSolvers` to potentially implement some subset of its operations via calls to the package `ARPACK`. Best

**Comment by ViralBShah** [comment]
I thought that eventually, the ARPACK wrappers would move to ARPACK.jl, and the rest of the higher level functionality would move to IterativeSolvers.jl.

**Comment by ararslan** [comment]
That also seems like a good option.

File: test/choosetests.jl (hunk 1 above)
Comment (ViralBShah) [reply test/choosetests.jl]: Yes, the shift and invert solvers need SuiteSparse.

**Comment by ViralBShah** [comment]
Let's go ahead with `IterativeEigenSolvers` for now, and get this done.

**Comment by ararslan** [comment]
Sounds great. Thanks all for the feedback.

**Comment by ararslan** [comment]
CI failures are unrelated (32-bit Windows couldn't connect, macOS couldn't load OpenBLAS). Unless there are any further comments, I plan to merge this soon.

**Review by Sacha0 — APPROVED** [review]
Looks great! Thanks `@ararslan`! :)

**Pull request merged**